dvc/dvcfile.py

```diff
@@ -0,0 +1,255 @@
+        """Dumps given stage appropriately in the dvcfile."""
+        self.dump_single_stage(stage)
+
+    def dump_single_stage(self, stage):
+        self.check_dvc_filename(self.path)
+
+        logger.debug(
+            "Saving information to '{file}'.".format(file=relpath(self.path))
+        )
+        state = stage.dumpd()
+
+        # When we load a stage we parse yaml with a fast parser, which strips
+        # off all the comments and formatting. To retain those on update we do
+        # a trick here:
+        # - reparse the same yaml text with a slow but smart ruamel yaml parser
+        # - apply changes to a returned structure
+        # - serialize it
+        if stage._stage_text is not None:
+            saved_state = parse_stage_for_update(stage._stage_text, self.path)
+            # Stage doesn't work with meta in any way, so .dumpd() doesn't
+            # have it. We simply copy it over.
+            if "meta" in saved_state:
+                state["meta"] = saved_state["meta"]
+            apply_diff(state, saved_state)
+            state = saved_state
+
+        dump_stage_file(self.path, state)
+
+        self.repo.scm.track_file(relpath(self.path))
+
+    def load(self):
+        """Loads single stage."""
+        from dvc.stage import Stage
+
+        # it raises the proper exceptions by priority:
+        # 1. when the file doesn't exists
+        # 2. filename is not a DVC-file
+        # 3. path doesn't represent a regular file
+        self.check_file_exists()
+        self.check_dvc_filename(self.path)
+        self.check_isfile()
+
+        with self.repo.tree.open(self.path) as fd:
+            stage_text = fd.read()
+        d = parse_stage(stage_text, self.path)
+
+        Dvcfile.validate(d, fname=relpath(self.path))
+        path = os.path.abspath(self.path)
+
+        stage = Stage(
+            repo=self.repo,
+            path=path,
+            wdir=os.path.abspath(
+                os.path.join(
+                    os.path.dirname(path), d.get(Stage.PARAM_WDIR, ".")
+                )
+            ),
+            cmd=d.get(Stage.PARAM_CMD),
+            md5=d.get(Stage.PARAM_MD5),
+            locked=d.get(Stage.PARAM_LOCKED, False),
+            tag=self.tag,
+            always_changed=d.get(Stage.PARAM_ALWAYS_CHANGED, False),
+            # We store stage text to apply updates to the same structure
+            stage_text=stage_text,
+        )
+
+        stage.deps = dependency.loadd_from(
+            stage, d.get(Stage.PARAM_DEPS) or []
+        )
+        stage.outs = output.loadd_from(stage, d.get(Stage.PARAM_OUTS) or [])
+
+        return stage
+
+    @staticmethod
+    def validate(d, fname=None):
+        from dvc.stage.schema import SINGLE_STAGE_SCHEMA
+
+        try:
+            SINGLE_STAGE_SCHEMA(d)
+        except MultipleInvalid as exc:
+            raise StageFileFormatError(fname, exc)
+
+    @classmethod
+    def create_stage(cls, repo, accompany_outs=False, **kwargs):
```

Comment:
Pretty much the whole logic doesn't belong to dvcfile itself, but belongs to the Stage class. Seems like the only thing that `Dvcfile` should care about here is the filename.

Reply:
E.g. it should care about `if os.path.exist` at the bottom. So, for example, it could be `Dvcfile.create()` method, that would in turn call `Stage.create()` with all that meat and then would check for already existing dvcfile with a path like that.

Reply:
It needs to be somewhere in between of those two. `Dvcfile` should care about filename, if it's a multistage dvc, if stage already exists with the given name, etc.
Rest can be taken care by `Stage`, but it's not really linear at the time. :(

Reply:
@skshetry Yes, that was my point. Right now `create_stage` does too much with the Stage. It should pass to `Stage.create()` and deal with dvcfile file/existing dvcfile/etc.

Reply:
Took a closer look. Looks like `Dvcfile.create` should handle things like filename/wdir validation. E.g. things like `_stage_name` and `_check_stage_path` don't belong to `Stage` anymore, they should be handled in `Dvcfile`.

Reply:
Did a bit of refactor. Most of the things stays on `Stage::create()`. Only overwriting stuffs have been pulled to `Dvcfile`.